src/libraries/Common/src/Roslyn/MetadataHelpers.cs

```diff
@@ -0,0 +1,1034 @@
+                                typeNameBuilder.Append(_input.Substring(start, _offset - start));
+                                return;
+                            }
+
+                            isMultiDimensionalIfRankOne = true;
+                            break;
+
+                        default:
+                            // Error case, process as regular characters
+                            Advance();
+                            typeNameBuilder.Append(_input.Substring(start, _offset - start));
+                            return;
+                    }
+                }
+
+                // Error case, process as regular characters
+                typeNameBuilder.Append(_input.Substring(start, _offset - start));
+            }
+        }
+
+        private static readonly string[] s_aritySuffixesOneToNine = { "`1", "`2", "`3", "`4", "`5", "`6", "`7", "`8", "`9" };
+
+        internal static string GetAritySuffix(int arity)
+        {
+            Debug.Assert(arity > 0);
+            return (arity <= 9) ? s_aritySuffixesOneToNine[arity - 1] : string.Concat(GenericTypeNameManglingString, arity.ToString(CultureInfo.InvariantCulture));
+        }
+
+        internal static string ComposeAritySuffixedMetadataName(string name, int arity)
+        {
+            return arity == 0 ? name : name + GetAritySuffix(arity);
+        }
+
+#endif
+
+        internal static int InferTypeArityFromMetadataName(string emittedTypeName)
+        {
+            return InferTypeArityFromMetadataName(emittedTypeName, out _);
+        }
+
+        private static short InferTypeArityFromMetadataName(string emittedTypeName, out int suffixStartsAt)
+        {
+            Debug.Assert(emittedTypeName != null, "NULL actual name unexpected!!!");
+            int emittedTypeNameLength = emittedTypeName.Length;
+
+            int indexOfManglingChar;
+            for (indexOfManglingChar = emittedTypeNameLength; indexOfManglingChar >= 1; indexOfManglingChar--)
+            {
+                if (emittedTypeName[indexOfManglingChar - 1] == GenericTypeNameManglingChar)
+                {
+                    break;
+                }
+            }
+
+            if (indexOfManglingChar < 2 ||
+               (emittedTypeNameLength - indexOfManglingChar) == 0 ||
+               emittedTypeNameLength - indexOfManglingChar > MaxStringLengthForParamSize)
+            {
+                suffixStartsAt = -1;
+                return 0;
+            }
+
+            // Given a name corresponding to <unmangledName>`<arity>,
+            // extract the arity.
+            string stringRepresentingArity = emittedTypeName.Substring(indexOfManglingChar);
+
+            int arity;
+            bool nonNumericCharFound = !int.TryParse(stringRepresentingArity, NumberStyles.None, CultureInfo.InvariantCulture, out arity);
+
+            if (nonNumericCharFound || arity < 0 || arity > short.MaxValue ||
+                stringRepresentingArity != arity.ToString())
```

Comment:
Why are we .ToString()'ing the arity again?  Is this to handle cases where stringRepresentingArity had trailing nulls?